extensions/docker/deployment/src/main/java/io/quarkus/docker/deployment/DockerBuildStep.java

```diff
@@ -0,0 +1,123 @@
+
+package io.quarkus.docker.deployment;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ApplicationInfoBuildItem;
+import io.quarkus.deployment.pkg.builditem.ContainerImageBuildItem;
+import io.quarkus.deployment.pkg.builditem.ContainerImageResultBuildItem;
+import io.quarkus.deployment.pkg.builditem.JarBuildItem;
+import io.quarkus.deployment.pkg.builditem.NativeImageBuildItem;
+import io.quarkus.deployment.pkg.builditem.OutputTargetBuildItem;
+import io.quarkus.deployment.pkg.steps.NativeBuild;
+import io.quarkus.deployment.util.ExecUtil;
+import io.quarkus.deployment.util.ImageUtil;
+
+public class DockerBuildStep {
+
+    private static final Logger log = Logger.getLogger(DockerBuildStep.class);
+    private static final String DOCKERFILE_JVM = "Dockerfile.jvm";
+    private static final String DOCKERFILE_NATIVE = "Dockerfile.native";
+
+    @BuildStep(onlyIf = DockerBuild.class, onlyIfNot = NativeBuild.class)
+    public ContainerImageResultBuildItem dockerBuildFromJar(ApplicationInfoBuildItem app,
+            OutputTargetBuildItem out,
+            Optional<ContainerImageBuildItem> dockerImage,
+            JarBuildItem artifact) {
+        log.info("Building docker image for jar.");
+        String image = dockerImage.map(d -> d.getImage()).orElse(app.getName() + ":" + app.getVersion());
+        ImageIdReader reader = new ImageIdReader();
+        Path dockerFile = extractDockerfile(DOCKERFILE_JVM);
+        ExecUtil.exec(out.getOutputDirectory().toFile(),
+                reader,
+                "docker", "build",
+                "-f",
+                dockerFile.resolve(DOCKERFILE_JVM).toAbsolutePath().toString(),
+                "-t", image,
+                out.getOutputDirectory().toAbsolutePath().toString());
+
+        return new ContainerImageResultBuildItem(reader.getImageId(), ImageUtil.getRepository(image), ImageUtil.getTag(image));
+    }
+
+    @BuildStep(onlyIf = { DockerBuild.class, NativeBuild.class })
+    public ContainerImageResultBuildItem dockerBuildFromNativeImage(ApplicationInfoBuildItem app,
+            OutputTargetBuildItem out,
+            Optional<ContainerImageBuildItem> dockerImage,
+            NativeImageBuildItem nativeImage) {
+        log.info("Building docker image for native image.");
+        String image = dockerImage.map(d -> d.getImage()).orElse(app.getName() + ":" + app.getVersion() + "-native");
+        Path dockerFile = extractDockerfile(DOCKERFILE_NATIVE);
```

Comment:
Not a priority, but it would be nice to allow the user to package a dockerfile in their app and use that to provide a way to customise it.

Reply:
Agree, but this needs discussion...

My initial intention was to use the provided Dockerfiles. The user would be able to customize them.

However, in order to use them I had to introduce the `ModuleRootBuildItem` so the container extension can read the provided Dockerfiles. In earlier comments it was suggested that we want to avoid using something like the `ModuleRootBuildItem`.

Reply:
One option here would be to put the dockerfiles in src/main/resources, so they can be read out of the class path.